xy-heatmap-category/parseData.js

```diff
@@ -0,0 +1,68 @@
+/**
+ * General data munging functionality
+ */
+
+import * as d3 from 'd3';
+import loadData from '@financial-times/load-data';
+
+/**
+ * Parses data file and returns structured data
+ * @param  {String} url Path to CSV/TSV/JSON file
+ * @return {Object}     Object containing series names, value extent and raw data object
+ */
+export function load(url, options) { // eslint-disable-line
+    return loadData(url).then((result) => {
+        const data = result.data ? result.data : result;
+        const seriesNames = getSeriesNames(data.columns);
+
+        const groupNames = data.map(d => d.name).filter(d => d); // create an array of the group names
+
+        const catNames = getCatNames(data, seriesNames);
+
+        // Buid the dataset for plotting
+        const plotData = data.map(d => ({
+            name: d.name,
+            groups: getGroups(seriesNames, d),
+        }));
+
+        return {
+            seriesNames,
+            plotData,
+            data,
+            groupNames,
+            catNames,
+        };
+    });
+}
+
+
+// a function that returns the columns headers from the top of the dataset, excluding specified
+function getSeriesNames(columns) {
+    const exclude = ['name']; // adjust column headings to match your dataset
+    return columns.filter(d => (exclude.indexOf(d) === -1));
+}
+
+function getGroups(seriesNames, el) {
+    return seriesNames.map(name => ({
+        name,
+        value: el[name],
+    }));
+}
+
+function getCatNames(data, seriesNames) {
+    // look at all the values
+    const allVals = [];
+    // extract unique values used in heatmap
+    data.map((d) => {
+        seriesNames.forEach((e) => {
+            allVals.push(d[e]);
+        });
+    });
```

Comment:
Lines 53-60 can be rewritten as:

```js
const allVals = data.reduce((acc, cur) => acc.concat(seriesNames.map(e => d[e])), []);
```

Reply:
How does that work? As d is not defined. Also throws an undefined error on cur

Reply:
@sdbernard Whups, sorry — try:

```
const allVals = data.reduce((acc, cur) => acc.concat(seriesNames.map(d => cur[d])), []);
```